scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +444,237 @@
+    true
+    (catch clojure.lang.ExceptionInfo e
+      false)))
 
-  (timers/time!
-    handle-resource-offer!-duration
-    (try
-      (loop [] ;; This loop is for compare-and-set! below
-        (let [scheduler-contents @pending-jobs
-              db (db conn)
-              considerable (filter (fn [job]
-                                     (and (try
-                                            (d/invoke db :job/allowed-to-start? db (:db/id job))
-                                            true
-                                            (catch clojure.lang.ExceptionInfo e
-                                              false))
-                                          ;; ensure it always tries new hosts
-                                          (not-any? #(= % (:hostname offer))
-                                                    (->> (:job/instance job)
-                                                         (filter #(false? (:instance/preempted? %)))
-                                                         (map :instance/hostname)))))
-                                   scheduler-contents)
-              matches (timers/time!
-                        handle-resource-offer!-match-duration
-                        (match-offer-to-schedule considerable offer))
-              ;; We know that matches always form a prefix of the scheduler's contents
-              new-scheduler-contents (remove (set matches) scheduler-contents)
-              first-considerable-resources (-> considerable first util/job-ent->resources)
-              match-resource-requirements (util/sum-resources-of-jobs matches)]
-          (reset! front-of-job-queue-mem-atom
-                  (or (:mem first-considerable-resources) 0))
-          (reset! front-of-job-queue-cpus-atom
-                  (or (:cpus first-considerable-resources) 0))
-          (cond
-            (empty? matches) (decline-offers driver (:ids offer))
-            (not (compare-and-set! pending-jobs scheduler-contents new-scheduler-contents))
-            (do (log/info "Pending job atom contention encountered")
+(defn ids-of-backfilled-instances
+  "Returns a list of the ids of running, backfilled instances of the given job"
+  [job]
+  (->> (:job/instance job)
+       (filter #(and (= :instance.status/running (:instance/status %))
+                     (:instance/backfilled? %)))
+       (map :db/id)))
+
+(defn process-matches-for-backfill
+  "This computes some sets:
+   
+   fully-processed: this is a set of job uuids that should be removed from the list scheduler jobs. These are not backfilled
+   upgrade-backfill: this is a set of instance datomic ids that should be upgraded to non-backfill
+   backfill-jobs: this is a set of job uuids for jobs that were matched, but are in the backfill QoS class
+   "
+  [scheduler-contents head-of-considerable matched-jobs]
+  (let [matched-job-uuids (set (mapv :job/uuid matched-jobs))
+        head-uuid (:job/uuid head-of-considerable)]
```

Comment:
In this chain of comments you dismissed one of the big problems you feared, but that leaves this:

> A bunch of jobs are scheduled, one of the jobs, J, will not get out of state unknown. All of the jobs ahead of it in the queue will be removed from pending on the next rank jobs, however, J will remain since it is still "waiting". J is now at the head of the queue but since it is not running, it will cause all jobs after it to be considered backfilled.

Well... if this job is left in state "waiting" (let's say it stays there for a long time because Mesos doesn't receive the message, or something), wouldn't it possibly get matched again on a subsequent pass to Fenzo?  If it gets matched, it wouldn't cause all jobs behind it to be backfilled.